crates/sui-core/src/authority/authority_per_epoch_store.rs

```diff
@@ -80,6 +86,16 @@ pub struct ExecutionIndicesWithHash {
     pub hash: u64,
 }
 
+// Data related to VM and Move execution and type layout
+pub struct ExecutionComponents {
+    /// Move native functions that are available to invoke
+    pub(crate) native_functions: NativeFunctionTable,
+    pub(crate) move_vm: Arc<MoveVM>,
+    // TODO: use strategies (e.g. LRU?) to constraint memory usage
+    pub(crate) module_cache: Arc<SyncModuleCache<ResolverWrapper<AuthorityStore>>>,
+    metrics: Arc<ResolverMetrics>,
```

Comment:
this is a bit of an annoyance and we need to keep `metrics` both here and inside `module_cache`.
The problem is that we cannot fetch it again from the `module_cache` which is a Move concept and we would have the change that. So we keep it in both place and use this when restarting a vm and cache across epochs.
Work on the loader is supposed to fi a lot of those vm usages, so a good solution may happen at a later point in time.
If anybody has an idea on how to make this better, please let me know

Reply:
I think this is a necessary evil -- it's quite common to store metrics like this (for prometheus) in an Arc and clone that into all the necessary places.